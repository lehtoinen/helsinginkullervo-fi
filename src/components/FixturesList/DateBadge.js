import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import locale from 'moment/locale/fi'; // eslint-disable-line no-unused-vars

import styles from './DateBadge.module.css';

const DateBadge = ({ date, isPast }) => {
  moment.locale('fi');
  const eventDate = moment(date);
  const weekday = eventDate.format('dd');
  const dateStr = eventDate.format('DD.MM');

  return (
    <div className={`${styles.root} ${isPast ? styles.completed : null}`}>
      <div className={styles.day}>
        <div className={styles.weekday}>{weekday}</div>
        <div className={styles.date}>{dateStr}</div>
      </div>
    </div>
  );
};

DateBadge.propTypes = {
  date: PropTypes.instanceOf(Date).isRequired,
  isPast: PropTypes.bool.isRequired,
};

export default DateBadge;
