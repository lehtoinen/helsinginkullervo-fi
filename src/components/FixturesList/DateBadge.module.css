.root {
  padding-top: 1.25em;
  margin-top: 1.25em;
  border-top: 1px dashed var(--color-complimentary);
}

.root.completed {
  opacity: 0.5;
}

.day {
  text-align: center;
  line-height: 1.15;
  position: absolute;
}

.weekday {
  font-family: Nunito, sans-serif;
  font-weight: 700;
  font-size: 1.45em;
  text-transform: uppercase;
  color: var(--color-base-dark);
}

.date {
  font-size: 0.8em;
  color: var(--color-grey);
  margin-top: 0.1em;
}
