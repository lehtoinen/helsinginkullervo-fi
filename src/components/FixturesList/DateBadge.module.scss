@import '../../assets/scss/colors';

.root {
  padding-top: 1.25em;
  margin-top: 1.25em;
  border-top: 1px dashed $color-complimentary;

  &.completed {
    opacity: 0.5;
  }

  .day {
    text-align: center;
    margin-right: 0.85em;
    line-height: 0.975;
    position: absolute;

    .weekday {
      font-size: 1.65em;
      font-weight: bold;
      text-transform: uppercase;
      color: $color-base-dark;
    }

    .date {
      font-size: 0.8em;
      color: $color-grey;
    }
  }
}