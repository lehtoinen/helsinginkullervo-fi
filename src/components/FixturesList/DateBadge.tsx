import React from 'react';
import { format } from 'date-fns';
import { fi } from 'date-fns/locale';

import * as styles from './DateBadge.module.css';

type Props = {
  date: Date;
  isPast: boolean;
};

const DateBadge = ({ date, isPast }: Props) => {
  const weekday = format(date, 'EEEEEE', { locale: fi });
  const dateStr = format(date, 'dd.MM');

  return (
    <div className={`${styles.root} ${isPast ? styles.completed : null}`}>
      <div className={styles.day}>
        <div className={styles.weekday}>{weekday}</div>
        <div className={styles.date}>{dateStr}</div>
      </div>
    </div>
  );
};

export default DateBadge;
