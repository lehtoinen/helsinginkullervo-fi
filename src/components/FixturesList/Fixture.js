import React from 'react';
import PropTypes from 'prop-types';

import styles from './Fixture.module.scss';

const Fixture = props => {
  const time = props.time.substr(0, props.time.length - 3);
  let winner;
  if (props.isCompleted) winner = props.homeScore - props.awayScore;

  return (
    <div
      className={`${styles.fixture} ${
        props.isCompleted ? styles.completed : null
      }`}
    >
      <div className={styles.subject}>
        <span className={styles.competition}> {props.competition}: </span>
        <span className={winner > 0 ? styles.winner : null}>
          {props.homeTeam} {props.homeScore}
        </span>{' '}
        -{' '}
        <span className={winner < 0 ? styles.winner : null}>
          {props.awayScore} {props.awayTeam}
        </span>
      </div>
      <div className={styles.info}>
        klo {time} <span className={styles.location}>{props.venue}</span>
      </div>
    </div>
  );
};

Fixture.propTypes = {
  time: PropTypes.string.isRequired,
  competition: PropTypes.string.isRequired,
  homeTeam: PropTypes.string.isRequired,
  awayTeam: PropTypes.string.isRequired,
  isCompleted: PropTypes.bool.isRequired,
  venue: PropTypes.string,
  homeScore: PropTypes.string,
  awayScore: PropTypes.string,
};

Fixture.defaultProps = {
  venue: null,
  homeScore: null,
  awayScore: null,
};

export default Fixture;
