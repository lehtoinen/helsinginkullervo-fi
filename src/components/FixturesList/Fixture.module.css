.fixture {
  padding-left: 3.5em;
  line-height: 1.25;
  margin-bottom: 1.35em;
  color: var(--color-black);
}

.fixture.completed {
  opacity: 0.5;
}

.winner {
  font-weight: bold;
}

.subject {
  font-size: 1.1em;
  font-style: italic;
}

.info {
  font-size: 0.8em;
  margin-top: 0.4em;
  text-transform: uppercase;
  letter-spacing: 0.05em;
}

.competition {
  font-style: normal;
}

.location {
  color: var(--color-grey);
  border-left: 1px solid #e6e6e6;
  padding-left: 0.75em;
  margin: 0.4em;
}
