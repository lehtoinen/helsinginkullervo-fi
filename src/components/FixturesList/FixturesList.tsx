import React, { Fragment } from 'react';

import { Fixture } from '../../types';

import FixtureListEntry from './FixtureListEntry';
import DateBadge from './DateBadge';

type Props = {
  fixtures: Fixture[];
};

const FixturesList = ({ fixtures = [] }: Props) => {
  let currentDate: string;

  return (
    <>
      {fixtures.map((fixture) => {
        const addDateBadge = currentDate !== fixture.date;
        currentDate = fixture.date;

        return (
          <Fragment
            key={`${fixture.homeTeam}${fixture.awayTeam}${fixture.date}`}
          >
            {addDateBadge && (
              <DateBadge
                date={new Date(fixture.date)}
                isPast={fixture.isCompleted}
              />
            )}
            <FixtureListEntry {...fixture} />
          </Fragment>
        );
      })}
    </>
  );
};

export default FixturesList;
