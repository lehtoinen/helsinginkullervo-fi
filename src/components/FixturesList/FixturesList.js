import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { updateFixtureFilters } from '../../state/actions';

import FixturesFilter from './FixturesFilter';
import Fixture from './Fixture';
import DateBadge from './DateBadge';

export class FixturesList extends React.Component {
  static parseCompetitions(fixtures) {
    const competitions = [];
    fixtures.forEach(fixture => {
      const { competition } = fixture;
      if (!competitions.includes(competition)) {
        competitions.push(competition);
      }
    });
    return competitions.sort();
  }

  static filterFixtures(fixtures, filters) {
    let filtered = fixtures;

    // filter fixtures by completion status
    if (filters.upcoming && filters.upcoming.length) {
      filtered = filtered.slice(
        filtered.findIndex(fixture => !fixture.isCompleted)
      );
    }

    // filter fixtures by competition
    if (filters.competitions && filters.competitions.length) {
      filtered = filtered.filter(fixture =>
        filters.competitions.includes(fixture.competition)
      );
    }

    return filtered;
  }

  onChangeFilter(type, value) {
    const filterValues = this.props.filters[type]
      ? this.props.filters[type].slice(0)
      : [];
    const ind = filterValues.indexOf(value);
    if (ind === -1) {
      filterValues.push(value);
    } else {
      filterValues.splice(ind, 1);
    }

    this.props.updateFixtureFilters({ [type]: filterValues });
  }

  render() {
    if (!this.props.fixtures || this.props.fixtures.length < 1) {
      return null;
    }

    const { filters } = this.props;
    const competitions = FixturesList.parseCompetitions(this.props.fixtures);
    const fixtures = FixturesList.filterFixtures(this.props.fixtures, filters);

    let currentDate;

    return (
      <div>
        <Fragment>
          <FixturesFilter
            options={competitions}
            selected={filters.competitions}
            onChange={option => this.onChangeFilter('competitions', option)}
          />
          <FixturesFilter
            options={['Näytä vain tulevat ottelut']}
            selected={filters.upcoming}
            onChange={option => this.onChangeFilter('upcoming', option)}
          />
        </Fragment>
        {fixtures.map(fixture => {
          const addDateBadge = currentDate !== fixture.date;
          currentDate = fixture.date;

          return (
            <Fragment key={`${fixture.competition}${fixture.date}`}>
              {addDateBadge && (
                <DateBadge
                  date={new Date(fixture.date)}
                  isPast={fixture.isCompleted}
                />
              )}
              <Fixture {...fixture} />
            </Fragment>
          );
        })}
      </div>
    );
  }
}

FixturesList.propTypes = {
  updateFixtureFilters: PropTypes.func.isRequired,
  filters: PropTypes.objectOf(PropTypes.array).isRequired,
  fixtures: PropTypes.arrayOf(
    PropTypes.shape({
      awayScore: PropTypes.string,
      awayTeam: PropTypes.string,
      competition: PropTypes.string,
      date: PropTypes.string,
      homeScore: PropTypes.string,
      homeTeam: PropTypes.string,
      isCompleted: PropTypes.bool,
      source: PropTypes.string,
      time: PropTypes.string,
      timecode: PropTypes.string,
      venue: PropTypes.string,
    })
  ),
};

FixturesList.defaultProps = {
  fixtures: [],
};

const mapStateToProps = state => ({
  filters: state.fixtureFilters,
});

const mapDispatchToProps = dispatch => ({
  updateFixtureFilters: filters => dispatch(updateFixtureFilters(filters)),
});

export default connect(mapStateToProps, mapDispatchToProps)(FixturesList);
