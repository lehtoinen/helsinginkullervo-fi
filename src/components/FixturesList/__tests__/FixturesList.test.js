import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { FixturesList } from '../FixturesList';

configure({ adapter: new Adapter() });

describe('FixturesList', () => {
  let component;

  const fetchFixtures = jest.fn();
  const updateFixtureFilters = jest.fn();
  const isLoading = false;
  const fixturesURL = 'foobarh';
  const fixtures = [];
  const filters = {};

  beforeEach(() => {
    // pass the mock function as the login prop
    component = shallow(
      <FixturesList
        fetchFixtures={fetchFixtures}
        updateFixtureFilters={updateFixtureFilters}
        isLoading={isLoading}
        fixturesURL={fixturesURL}
        fixtures={fixtures}
        filters={filters}
      />
    );
  });

  test('calls the fetchFixtures function', () => {
    expect(fetchFixtures.mock.calls.length).toBe(1);
  });

  test('renders', () => {
    expect(component).toMatchSnapshot();
  });
});
