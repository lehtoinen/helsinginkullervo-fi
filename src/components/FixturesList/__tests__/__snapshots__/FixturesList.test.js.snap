// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FixturesList renders 1`] = `
<div>
  <React.Fragment>
    <FixturesFilter
      onChange={[Function]}
      options={
        Array [
          "Regions Cup",
        ]
      }
      selected={Array []}
    />
    <FixturesFilter
      onChange={[Function]}
      options={
        Array [
          "Näytä vain tulevat ottelut",
        ]
      }
      selected={Array []}
    />
  </React.Fragment>
  <React.Fragment
    key="Regions Cup2018-04-15"
  >
    <DateBadge
      date={2018-04-15T00:00:00.000Z}
      isPast={true}
    />
    <Fixture
      awayScore="0"
      awayTeam="Kullervo/Überkleber"
      competition="Regions Cup"
      date="2018-04-15"
      homeScore="8"
      homeTeam="RiPS"
      isCompleted={true}
      time="15:00:00"
      timecode="2018-04-15-15:00:00"
      venue="Riihimäki Keskuskenttä TN"
    />
  </React.Fragment>
</div>
`;

exports[`FixturesList renders with undefined fixtures 1`] = `""`;
